feat(pack): add built-in plugin manager `vim.pack` (#34009)
neovim/neovim: runtime/doc/lua.txt

      • {path}  (`string?`) path to reset

==============================================================================
Lua module: vim.pack                                                *vim.pack*

WORK IN PROGRESS built-in plugin manager! Early testing of existing features
is appreciated, but expect breaking changes without notice.

Manages plugins only in a dedicated |packages| *vim.pack-directory*:
`$XDG_DATA_HOME/nvim/site/pack/core/opt`. Plugin's subdirectory name matches
plugin's name in specification. It is assumed that all plugins in the
directory are managed exclusively by `vim.pack`.

Uses Git to manage plugins and requires present `git` executable of at least
version 2.36. Target plugins should be Git repositories with versions as named
tags following semver convention `v<major>.<minor>.<patch>`.

Example workflows ~

Basic install and management:
• Add |vim.pack.add()| call(s) to 'init.lua': >lua

    vim.pack.add({
      -- Install as "plugin1"/"plugin2" and use greatest available version
      'https://github.com/user/plugin1',

>       { source = 'https://github.com/user/plugin2' },

Reviewer: This comment and the way these two lines are grouped together is confusing (I had to read this 3 times and I'm still not 100% confident I'm understanding it correctly).

IIUC, this installs two different plugins, `github.com/user/plugin1` and `github.com/user/plugin2`. The 1st example is demonstrating the "shortcut" method of installing a plugin and the 2nd example is showing the more verbose version.

The single comment and grouping of these lines makes it seem like this is installing a single plugin under a directory `plugin1/plugin2`.

If my understanding is correct, I think formatting it like this would make it clearer:

```
-- Install "plugin1" and use greatest available version.
'https://github.com/user/plugin1',

-- Same as above, but using a full table, which allows setting other options.
{ source = 'https://github.com/user/plugin1' },
```

>       -- Specify plugin's name
>       { source = 'https://github.com/user/generic-name', name = 'plugin3' },

Reviewer: Maybe clarify here that if `name` is omitted it's inferred from the URL?

```suggestion
      -- Specify plugin's name (here the plugin will be called "plugin3"
      -- instead of "generic-name")
      { source = 'https://github.com/user/generic-name', name = 'plugin3' },
```

      -- Specify version to follow during install and update
      {
        source = 'https://github.com/user/plugin4',

>         version = vim.version.range('1.0'), -- Greatest version matching range

Reviewer: I'm not sure what "Greatest version matching range" means. Maybe call this the "version constraint" instead (with a link to the docs).

I had to read the `vim.version.range` docs to understand exactly what `vim.version.range('1.0')` means. This matches `1.0.0` and `1.0.1`, but not `1.1`, for example.

```suggestion
        version = vim.version.range('1.0'), -- Version constraint. See |vim.version.range()|
```

Author: > I'm not sure what "Greatest version matching range" means. Maybe call this the "version constraint" instead (with a link to the docs).

In `vim.version` terminology it is "last", which I don't find correct/intuitive. It is the largest/biggest/greatest available version that matches any constraint specified via `vim.version.range()`. In this particular case it is indeed larger than or equal `1.0.0` but strictly less than `1.1.0`.

Yes, linking to `vim.version.range()` makes sense.

      },
      {
        source = 'https://github.com/user/plugin5',
        version = 'main', -- Git branch, tag, or commit hash
      },
    })

    -- Plugin's code can be used directly after `add()`
    plugin1 = require('plugin1')
<
• Restart Nvim. Plugins that were not yet installed will be available on disk
  in target state after `add()` call.
• To update all plugins with new changes:
  • Execute |vim.pack.update()|. This will download updates from source and
    show confirmation buffer in a separate tabpage.
  • Review changes. To confirm all updates execute |:write|. To discard

>     updates execute |:quit|.

Reviewer: Will this work outside of init.lua, for those of us who still use init.vim? For example, can I use `plugin/pack.lua` or even `lua/pack.lua` (and execute manually with `:lua require('pack')` when I want to modify my plugins)?

One reason I like using `start/` is that I don't have to dedicate any source code to loading plugins. I literally just `git clone` them to the right location and let Nvim handle the rest. Since it sounds like `vim.pack` puts everything in `opt`, I'm assuming that is not a use case that will be supported and I will have to use `vim.pack.add` to load plugins out of `opt`?

Reviewer 2: > that I don't have to dedicate any source code to loading plugins

If you use `vim.pack` wouldn't you then have source code for loading plugins regardless of whether it uses `opt` or `start` internally?

And if you are thinking from a perf pov, using a plugin manager with `opt` will be more efficient than using `start` since you can avoid an unnecessary directory path scan as the package manager knows where to look for plugins.

`start` really only makes sense if you do not want any plugin knowledge in your `init.lua`, and manage them externally, in which case a built-in package manager has no place. Unless I'm missing something?

Reviewer: I was a bit careless with my words/phrasing. Yes there will of course always be some code to specify the list of plugins which should be installed (even in the `start` world). I was referring to requiring code to just enable/load the plugin.

But I'm also speaking more from personal preference here. I think in regard to the builtin plugin manager, this design is perfectly fine. And the small minority of users still using init.vim instead of init.lua can use a Lua heredoc or something.

So consider this comment resolved.

Author: > Will this work outside of init.lua, for those of us who still use init.vim? For example, can I use `plugin/pack.lua` or even `lua/pack.lua` (and execute manually with `:lua require('pack')` when I want to modify my plugins)?

This should work as any "regular" `vim.xxx()` function, even via `:lua vim.pack.add(...)`.

And yes, 'start/' vs 'opt/' shouldn't make a difference here if they both are required to be listed in the config somewhere.

Switch plugin's version:
• Update 'init.lua' for plugin to have desired `version`. Let's say, plugin
  named 'plugin1' has changed from default `vim.version.range('*')` to `main`.

> • Restartn Nvim. This has not yet changed plugin's actual state on disk.

Reviewer: ```suggestion
• Restart Nvim. The plugin's actual state on disk is not yet changed.
```

• Execute `vim.pack.update({ 'plugin1' }, { offline = true })`. Using
  `offline` is optional to not download updates from source.

> • Review changes and either confirm or discard them. If discard, make sure to
>   revert changes in 'init.lua' also.

Reviewer: ```suggestion
• Review changes and either confirm or discard them. If discarded, revert any changes in 'init.lua' as well or you will be prompted again next time you run |vim.pack.update()|.
```

Freeze plugin from being updated:
• Update 'init.lua' for plugin to have `version = 'HEAD'`.

> • Restart Nvim.

Reviewer: I find it pretty counter-intuitive that `HEAD` implies freezing the plugin. In Git parlance, `HEAD` is the current tip, which is always changing.

Author: The logic is as follows. The `version` describes what target state to use when updating plugin. If target state is always "HEAD" (i.e. current), it means that updating will always "update" to the current state. I.e. no new changes.

I think this name best works if `version` is named `checkout`.

Naming "FREEZE" instead of "HEAD" is possible, but it makes it a "reserved" word which would make it impossible to use branch named "FREEZE". A *very* unlikely scenario which might be worth the risk, of course.

Other plugin managers use a separate spec field for that, but I really don't think it is needed here.

Reviewer 2: At best `HEAD` is ambiguous here.

I think a separate field is warranted. Adding special keywords that can conflict in a namespace is a worse way of solving this and a separate field is much more intuitive and easier to understand.

Author: > I think a separate field is warranted. Adding special keywords that can conflict in a namespace is a worse way of solving this and a separate field is much more intuitive and easier to understand.

Sure, sounds good. My objective here is to have as simple spec as possible. There are precedents of having too fields in plugin spec bring more confusion than good.

Reviewer 2: Another idea would be to use another type like `version=false`? Isn't as easy to understand as a separate field but avoids the name clashing problem.

Author: I did go with `version = false` to freeze plugins. Is it enough to resolve this?

Reviewer: If someone wants to freeze/pin a plugin to an exact version can’t they just set the tag/commit hash? Do we need this extra field/value at all?

Author: Yes, that's a fair point. The special case of "freezing" a plugin is mostly due to:
- Having a recipe of "Just set `version = false` to prevent updates" is more straightforward than "Run `vim.pack.update({ 'plugin-name' }, { offline = true })`, yank the hash for current state, and set `version` to it".
- There is a similar functionality in other plugin managers (although with a separate field). So this concept should be familiar to users of other plugin managers.

As "freezing" the plugin from updates is more or less common operation, I decided to have this to reduce friction.

If both of the above points are not important enough for the added complexity, it can indeed be removed. I also "prepared" for it just now by having shorter hashes (`--abbrev-commit`) be reported as state in the `vim.pack.update()` output. Another reason for removing it might be that it is possible to add later, while removing this capability is less easy.

Author: After further discussion on Matrix, it seemed best to postpone the support for `version = false` for freezing plugin in favor of suggesting to use explicit commit hash. At least until there is a lockfile support, because:
- It will provide a more proper initial install of plugin (and not only checkout the default branch) if there is a decision to have `version = false` support.
- It will make easier to get the commit hash of current state if there will be no `version = false` (or separate `pin = true`) support.

Unfreeze plugin to start receiving updates:
• Update 'init.lua' for plugin to have `version` set to whichever version you
  want it to be updated.
• Restart Nvim.

Remove plugin from disk:
• Delete plugin's directory manually. Make sure its spec is not included in

>   |vim.pack.add()| call in 'init.lua' or it will be reinstalled.

Reviewer: No API to delete plugins? Having to manually run `rm -rf ~/.local/share/nvim/site/pack/core/plugin-name` is a bit tedious. Not to mention it leaks an abstraction (the installation directory of plugins).

Why not provide `vim.pack.del("plugin-name")`?

Reviewer 3: I think the plan is to
1. remove/comment out `vim.pack.add()`
2. call `vim.clean()`

(Agreed that we absolutely need a way to clean up what we put in.)

Author: > Why not provide `vim.pack.del("plugin-name")`?

It is one of the possibility (listed in "Planned future work" of the first comment), which I am not against. There are/were others:

1. Provide a `vim.pack.clean()` to clean not used plugins. This might get tricky if there is no way to cherry-pick plugins, because there is no good way to see which plugins are intended to be removed and which are just not (yet) loaded. See [this comment](https://github.com/neovim/neovim/pull/34009#issuecomment-2880367306) and what it responds to. So I think it is only a good option if it provides an *interactive* confirmation buffer allowing to filter out plugins that don't need to be removed. This is what 'mini.deps' does, but I think it is too much for `vim.pack.add`.
1. Export a way to get plugin's path based on the name and suggest to use `vim.fn.delete()`. So it might be something like `vim.fn.delete(vim.pack.get('plugin-name').path)`.
1. Indeed provide `vim.pack.del('plugin-name')`. This can also trigger dedicated events (`PackDelPre` and `PackDel`) which can be used by plugins to do some kind of "cleanup".

At the moment, I think either option 2. or 3. are fine. They 3. indeed might be a bit cleaner.

Reviewer 2: > because there is no good way to see which plugins are intended to be removed and which are just not (yet) loaded

This is exactly why I added a `cond = function(load) end` spec field in pckr.nvim to defer loading a plugin. In order for a package manager to manage plugins, it needs to know what plugins it is managing. There isn't really a way around this.

Since you have the constraint that `vim.pack.add` _must_ load the plugin, I don't think we'll ever be able to robustly manage the removal of unused plugins.

Reviewer 2: > vim.fn.delete()

Or `vim.fs.rm()`

Author: > ... it needs to know what plugins it is managing. There isn't really a way around this.

Right now it is "all plugins inside '~/.local/share/nvim/site/pack/core/opt'". All of them can be updated, i.e. "managed".

> I don't think we'll ever be able to robustly manage the removal of unused plugins. That will just need to be an accepted limitation.

Yes, I did come to that conclusion also (as a result of comments in this PR). And I think this limitation is worthwhile as the idea of "just add `vim.pack.add()` line to config and assume plugin is present afterwards" is very convenient when used inside config.

Hence why I think providing only a manual way of `vim.pack.del()` is good enough. It is explicit and atomic operation, which can be more than just deleting a directory.

Author: There is now `vim.pack.del()` in this PR. It accepts array of names and removes those plugins from disk. Along with triggering dedicated events and some proper bookkeeping.
Is it enough to resolve this conversation?

                    format supported by `git clone` is allowed.
      • {name}?     (`string`) Name of plugin. Will be used as directory name.
                    Default: basename of `source`.
      • {version}?  (`string|vim.VersionRange`) Version to use for install and
                    updates. One of:
                    • Output of |vim.version.range()| to install the
                      greatest/last semver tag inside the range.
                    • Tag name.
                    • Branch name.
                    • "HEAD" to freeze current state from updates for already
                      installed plugin.
                    • Explicit commit hash.

                    Default is `vim.version.range('*')`, i.e. install the
                    greatest available version.

vim.pack.add({specs}, {opts})                                 *vim.pack.add()*
    Add plugin to current session
    • For each specification check that plugin exists on disk in
      |vim.pack-directory|:
      • If exists, do nothin in this step.
      • If doesn't exist, install it by downloading from `source` into `name`
        subdirectory (via `git clone`) and update state to match `version`

>         (via `git checkout`).

Reviewer: Does this download happen automatically? So if a user has `vim.pack.add` in their init.lua, as soon as they start Nvim it will automatically start downloading things?

I wonder if this should be a more explicit action. Nvim automatically downloading things without any user action feels a bit iffy (I don't think there's any precedent for that?)

Reviewer 3: Spell files? I would consider a `vim.pack.add` line as "explicitly opting in" here.

Reviewer: I've never downloaded a spell file in Vim, but reading the docs briefly it says 

>But if the "spellfile.vim" plugin is active it will offer you to download the spell file

In this case it's still requiring an explicit action (confirmation) from the user, not just immediately downloading.

>I would consider a vim.pack.add line as "explicitly opting in" here.

Users are required to have `vim.pack.add` in their init file to even load/enable the plugin. So I don't think it's acceptable to treat that as tacit approval of "yes please download things automatically".

It's not difficult to show a single confirmation dialog that says "Install the following plugins? [Y/n]"

Reviewer 3: No, I mean adding `vim.pack.add('thisplugin')` _is_ giving confirmation to "yes, please download _this plugin_". I don't want to have to deal with a second prompt for this. (Otherwise bootstrapping will be painful.)

> But if the "spellfile.vim" plugin is active it will offer you to download the spell file

Which it is by default, in Nvim (not Vim IIRC).

Reviewer: >Which it is by default, in Nvim (not Vim IIRC).

The plugin is enabled, but still "offers to download the spell file" (implying there is a prompt).

>No, I mean adding vim.pack.add('thisplugin') is giving confirmation to "yes, please download this plugin". I don't want to have to deal with a second prompt for this. (Otherwise bootstrapping will be painful.)

The bootstrapping situation is exactly the one I'm thinking of where I would want a prompt. If I'm on an existing machine and I want to add a new plugin, I agree that I should be able to run some command/call some function and add/install the new plugin all at once (that is my explicit opt-in).

But if I'm on a new machine, or perhaps reproducing a user's bug report which installs a bunch of arbitrary plugins, and I start Nvim for the first time, I absolutely want to explicitly say "yes go ahead and install the plugins after I see what you're going to install". The difference is that when you're starting Nvim for the first time on a new machine or reproducing some bug, you don't necessarily already know exactly what plugins are going to be downloaded.

Reviewer: >If I'm on an existing machine and I want to add a new plugin, I agree that I should be able to run some command/call some function and add/install the new plugin all at once (that is my explicit opt-in).

Being able to add a `vim.pack.add()` and then immediately call `:lua vim.pack.update()` to install the plugin that I've just added seems like it should satisfy this use case. Instead of having to add `vim.pack.add()` and then restart Nvim.

Reviewer 2: Pretty sure there is a prompt for spell files, and I also agree plugin installs should require a prompt.

Author: My thinking is what @clason described: putting explicit `vim.pack.add()` in the config should be enough of a confirmation already. 

> Being able to add a `vim.pack.add()` and then immediately call `:lua vim.pack.update()` to install the plugin that I've just added seems like it should satisfy this use case. Instead of having to add `vim.pack.add()` and then restart Nvim.

Plugin will be installed already after `vim.pack.add()`, no need to either restart Nvim or `vim.pack.update()`. So I don't really understand what this is about.

> But if I'm on a new machine, or perhaps reproducing a user's bug report which installs a bunch of arbitrary plugins, and I start Nvim for the first time, I absolutely want to explicitly say "yes go ahead and install the plugins after I see what you're going to install". The difference is that when you're starting Nvim for the first time on a new machine or reproducing some bug, you don't necessarily already know exactly what plugins are going to be downloaded.

Well, right now it is known (all plugins are listed), but if/when there is packspec support, this indeed won't be known 100% in advance.

I am open to adding some confirmation before installing. Some questions:
- Is something like `vim.fn.confirm('These plugins will be installed:\n...')` enough or should be more interactive (like allowing to specify which plugins to install)? I'd guess it should be enough.
- Should it be one confirmation per install (i.e. several times pressing `y`) or one confirm for all installs? I'd guess one for all plugins should be enough, otherwise it is a tedious amount of "y" during bootstrap.
- Should it be run only during startup (i.e. `vim.v.vim_did_enter == 0`) or for every install? I'd go with only during startup, because otherwise it should be a conscious decision (i.e. enough of confirmation already).

Reviewer 2: A single prompt for a list of plugins and done any time a plugin is loaded.

Reviewer: >Is something like vim.fn.confirm('These plugins will be installed:\n...') enough or should be more interactive 

Yes I think that's fine

>Should it be one confirmation per install (i.e. several times pressing y) or one confirm for all installs

One for all, just list all of the ones that are going to be installed (this can be done in a floating window or something if there are a lot)

>putting explicit vim.pack.add() in the config should be enough of a confirmation already.

There are cases where you DIDN'T put `vim.pack.add` in the config though (you are using a config that somebody else provided to reproduce a bug), or when you did the context/situation was different (maybe I put `vim.pack.add` in my config several months ago, but now I'm cloning onto a new machine and I want to review what's going to be downloaded before Nvim starts cloning a bunch of repos automatically).

> Plugin will be installed already after vim.pack.add(), no need to either restart Nvim or vim.pack.update(). So I don't really understand what this is about.

As a general design rule Nvim shouldn't be downloading things from the internet without some explicit consent/action from the user. If the user runs `vim.pack.add()` interactively then I agree that is sufficiently explicit. But if it's run programmatically (as part of startup in init.lua) I don't think it is, at least not as a default.

In Emacs, `use-package` has an `ensure` option that users can set to have the plugin automatically downloaded. That is another option we could explore (set either per-plugin as part of `vim.pack.add()` or globally as part of `vim.pack.config()`). But as a **default**, we should err on the side of requiring user action.

Author: There is now a confirm before any plugin(s) install. It uses `vim.fn.confirm()`, but during startup it is drawing issues (see #34088), so for now it is a bit hacky workaround with delayed `vim.print()` and manual processing of the key asked via `vim.fn.getcharstr()`. After #34088 is solved (or 'extui' becomes default, whichever is sooner), the `vim.fn.confirm()` can be used universally.

      present on disk. Execute |vim.pack.update()| (possibly with
      `{ offline = true }`) to synchronize.
    • Adding plugin second and more times during single session does nothing:
      only the data from the first adding is registered.

    Parameters: ~
      • {specs}  (`(string|vim.pack.Spec)[]`) List of plugin specifications.
                 String item is treated as `source`.
      • {opts}   (`table?`) A table with the following fields:
                 • {bang}? (`boolean`) Whether to execute `:packadd!` instead
                   of |:packadd|. Default `false`.

vim.pack.get()                                                *vim.pack.get()*
    Get data about all plugins managed by |vim.pack|

    Return: ~
        (`table[]`) A list of objects with the following fields:
        • {spec} (`vim.pack.SpecResolved`) A |vim.pack.Spec| with defaults
          made explicit.
        • {path} (`string`) Plugin's path on disk.
        • {was_added} (`boolean`) Whether plugin was added via
          |vim.pack.add()| in current session.

vim.pack.update({names}, {opts})                           *vim.pack.update()*
    Update plugins

>     • If not `offline`, download new changes from source.

Reviewer: I'm not sure I understand what this does if `offline` is false?

Author: Everything that is listed afterwards: infer the matching states from `version` field of plugins and either update (`force = true`) or show confirmation buffer (`force = false`, default). This is useful for changing the version of plugin on disk, as it doesn't require internet connection. Strictly speaking, having `offline` field is not needed if checking for upstream changes is considered a "cheap" enough operation.

Author: Oh, sorry, I misunderstood the question (thought it was about `offline = true`) :facepalm:
If `offline` is `false` (which is default), it downloads new changes from the source. Under the hood it is `git fetch` with some set of useful flags, which makes local copy of `origin` remote have the same state as on remote. After that, it proceeds with next steps.

Reviewer: > Oh, sorry, I misunderstood the question (thought it was about offline = true)

Ha that is actually what I meant, so you understood my question even though I asked it incorrectly.

Author: After thinking about it, maybe it is worth postponing the decision about whether `offline = true` is necessary until after there is a lockfile support. Similar to `version = false`.

The reason for this is that with lockfile it will be feasible to have the design of "`add()` ensures both presence and correct on disk spec (like `source` and `version`)" and "`update()` is only for downloading and applying changes from source". This simplifies the "change `version`" step at the cost of not being able to see beforehand and confirm/deny which changes will be applied: they would still be present in the log file for inspection, though.

The only downside right now would be that changing `version` of already installed plugin would require Internet connection. Which can be reasonable enough as a temporary (hopefully not for long) restriction.

Author: The `vim.pack.update()` now doesn't have `opts.offline` option. The decision about its usefulness is postponed until after there is lockfile support.

      • {path}  (`string?`) path to reset

==============================================================================

> Lua module: vim.pack                                                *vim.pack*

Author: I've added this to the list of future work.

WORK IN PROGRESS built-in plugin manager! Early testing of existing features
is appreciated, but expect breaking changes without notice.

Manages plugins only in a dedicated *vim.pack-directory* (see |packages|):
`$XDG_DATA_HOME/nvim/site/pack/core/opt`. Plugin's subdirectory name matches
plugin's name in specification. It is assumed that all plugins in the
directory are managed exclusively by `vim.pack`.

Uses Git to manage plugins and requires present `git` executable of at least
version 2.36. Target plugins should be Git repositories with versions as named
tags following semver convention `v<major>.<minor>.<patch>`.

Example workflows ~

Basic install and management:
• Add |vim.pack.add()| call(s) to 'init.lua': >lua

    vim.pack.add({
      -- Install "plugin1" and use default branch (usually `main` or `master`)

>       'https://github.com/user/plugin1',

Reviewer 2: I suggest we just error on such conflicts until someone actually wants this to be resolved.

>       'https://github.com/user/plugin1',

Author: > if `'https://foo.com/user/plugin1'` is also specified, how does vim.pack handle the name conflict ? Docs may want to mention that.

Right now it [normalizes all specs](https://github.com/neovim/neovim/blob/44b11296513974b1ddf7ef7629644be4e2f6be03/runtime/lua/vim/pack.lua#L274-L314) and throws error if there are actual conflicts in `src` or `version`. It will probably also be a good place to use `vim.version.intersect()` to gather several semver ranges as their intersection, which will be important after dependencies are handled somehow.

One thing to note is that it is done only inside a single `vim.pack.add()` call, not across different calls. In that case it silently assumes that 'plugin1' is already loaded and does nothing. There is room for improvement here, but it needs a broader discussion about what to do on such conflicts (allow overriding parts of specs or not, etc.).

      -- Same as above, but using a table (allows setting other options)

>       { source = 'https://github.com/user/plugin1' },

Reviewer 2: +1 `url`

• Update 'init.lua' for plugin to have `version` set to current commit hash.
  You can get it by running `vim.pack.update({ 'plugin-name' })` and yanking
  the word describing current state (looks like `abc12345`).
• |:restart|.

Unfreeze plugin to start receiving updates:
• Update 'init.lua' for plugin to have `version` set to whichever version you
  want it to be updated.
• |:restart|.

Remove plugins from disk:
• Use |vim.pack.del()| with a list of plugin names to remove. Make sure their
  specs are not included in |vim.pack.add()| call in 'init.lua' or they will
  be reinstalled.

Available events to hook into ~
• *PackInstallPre* - before trying to install plugin on disk.
• *PackInstall* - after installing plugin on disk in proper state. After
  |PackUpdatePre| and |PackUpdate|.
• *PackUpdatePre* - before trying to update plugin's state.
• *PackUpdate* - after plugin's state is updated.
• *PackDeletePre* - before removing plugin from disk.
• *PackDelete* - after removing plugin from disk.

*vim.pack.Spec*

    Fields: ~
      • {source}    (`string`) URI from which to install and pull updates. Any
                    format supported by `git clone` is allowed.

>       • {name}?     (`string`) Name of plugin. Will be used as directory name.

Reviewer 2: I'd suggest not as it's important to this design that the spec maintains ordering. If you make `name` a key then you won't be able to process plugins in the order they are defined.

vim.pack.add({specs}, {opts})                                 *vim.pack.add()*
    Add plugin to current session
    • For each specification check that plugin exists on disk in
      |vim.pack-directory|:
      • If exists, do nothin in this step.
      • If doesn't exist, install it by downloading from `source` into `name`
        subdirectory (via `git clone`) and update state to match `version`
        (via `git checkout`).
    • For each plugin execute |:packadd| making them reachable by Nvim.

    Notes:
    • Installation is done in parallel, but waits for all to finish before
      continuing next code execution.
    • If plugin is already present on disk, there are no checks about its
      present state. The specified `version` can be not the one actually
      present on disk. Execute |vim.pack.update()| to synchronize.
    • Adding plugin second and more times during single session does nothing:
      only the data from the first adding is registered.

    Parameters: ~
      • {specs}  (`(string|vim.pack.Spec)[]`) List of plugin specifications.
                 String item is treated as `source`.
      • {opts}   (`table?`) A table with the following fields:

>                  • {bang}? (`boolean`) Whether to execute `:packadd!` instead

Reviewer 4: might want to clarify that we mean `plugin/*` files, which is meaningfully different from 'plugin files'. Maybe 'Load files under `plugin/` and `ftdetect/` directories'?
